Yank/mpi.py

```diff
@@ -0,0 +1,421 @@
+            kwargs['broadcast_result'] = broadcast_result
+            kwargs['sync_nodes'] = sync_nodes
+            return run_single_node(rank, task, *args, **kwargs)
+        return _wrapper
+    return _on_single_node
+
+
+def distribute(task, distributed_args, *other_args, **kwargs):
+    """Map the task on a sequence of arguments to be executed on different nodes.
+
+    If MPI is not activated, this simply runs serially on this node. The
+    algorithm guarantees that each node will be assigned to the same job_id
+    (i.e. the index of the argument in distributed_args) every time.
+
+    Parameters
+    ----------
+    task : callable
+        The task to be distributed among nodes. The task will be called as
+        task(distributed_args[job_id], *other_args, **kwargs), so the parameter
+        to be distributed must the the first one.
+    distributed_args : iterable
+        The sequence of the parameters to distribute among nodes.
+    send_results_to : int or 'all', optional
+        If the string 'all', the result will be sent to all nodes. If an
+        int, the result will be send only to the node with rank send_results_to.
+        The return value of distribute depends on the value of this parameter
+        (default is None).
+    sync_nodes : bool, optional
+        If True, the nodes will be synchronized at the end of the
+        execution (i.e. the task will be blocking) even if the
+        result is not shared (default is False).
+
+    Other Parameters
+    ----------------
+    *other_args
+        Other parameters to pass to task beside the assigned distributed
+        parameters.
+    **kwargs
+        Keyword arguments to pass to task beside the assigned distributed
+        parameters.
+
+    Returns
+    -------
+    all_results : list
+        All the return values for all the arguments if the results where sent
+        to the node, or only the return values of the arguments processed by
+        this node otherwise.
+    arg_indices : list of int, optional
+        This is returned as part of a tuple (all_results, job_indices) only
+        if send_results_to is set to an int or None. In this case all_results[i]
+        is the return value of task(all_args[arg_indices[i]]).
+
+    Examples
+    --------
+    >>> def square(x):
+    ...     return x**2
+    >>> distribute(square, [1, 2, 3, 4], send_results_to='all')
+    [1, 4, 9, 16]
+
+    When send_results_to is not set to `all`, the return value include also
+    the indices of the arguments associated to the result.
+
+    >>> distribute(square, [1, 2, 3, 4], send_results_to=0)
+    ([1, 4, 9, 16], [0, 1, 2, 3])
+
+    """
+    send_results_to = kwargs.pop('send_results_to', None)
+    sync_nodes = kwargs.pop('sync_nodes', False)
+    mpicomm = get_mpicomm()
+    n_jobs = len(distributed_args)
+
+    # If MPI is not activated, just run serially.
+    if mpicomm is None:
+        logger.debug('Running {} serially.'.format(task.__name__))
```

Comment:
That's the case if you set `send_results_to=node_you_care_about` (or `'all'` in case all node needs it). In this particular case (i.e. without MPI) we check `send_results_to` only to determine the shape of the returned value. I probably need to document this better.